airbyte-integrations/connectors/destination-s3/src/main/java/io/airbyte/integrations/destination/s3/csv/S3CsvOutputFormatter.java

```diff
@@ -0,0 +1,269 @@
+    String outputFilename = getOutputFilename(uploadTimestamp);
+    String objectKey = String.join("/", outputPrefix, outputFilename);
+
+    LOGGER.info("Full S3 path for stream '{}': {}/{}", stream.getName(), config.getBucketName(),
+        objectKey);
+
+    // The stream transfer manager lets us greedily stream into S3. The native AWS SDK does not
+    // have support for streaming multipart uploads. The alternative is first writing the entire
+    // output to disk before loading into S3. This is not feasible with large input.
+    // Data is chunked into parts during the upload. A part is sent off to a queue to be uploaded
+    // once it has reached it's configured part size.
+    // Memory consumption is queue capacity * part size = 1 * 5 = 5 MB at current configurations.
+    this.uploadManager = new StreamTransferManager(config.getBucketName(), objectKey,
+        s3Client)
+            .numUploadThreads(1)
+            .queueCapacity(1)
+            .partSize(S3DestinationConstants.DEFAULT_PART_SIZE_MD);
+    // We only need one output stream as we only have one input stream. This is reasonably performant.
+    this.outputStream = uploadManager.getMultiPartOutputStreams().get(0);
+    this.csvPrinter = new CSVPrinter(new PrintWriter(outputStream, true, StandardCharsets.UTF_8),
+        CSVFormat.DEFAULT.withQuoteMode(QuoteMode.ALL)
+            .withHeader(getHeaders(sortedHeaders).toArray(new String[0])));
+  }
+
+  /**
+   * Get a sorted field list in the json object so that this object can be iterated through with a
+   * defined order later on.
+   */
+  static List<String> getSortedFields(JsonNode jsonSchema, S3CsvFormatConfig formatConfig) {
+    // When no flattening is needed, we do not care about iteration order.
+    if (formatConfig.getFlattening() == Flattening.NO) {
+      return Collections.emptyList();
+    }
+    if (formatConfig.getFlattening() == Flattening.ROOT_LEVEL) {
+      return MoreIterators.toList(jsonSchema.get("properties").fieldNames())
+          .stream().sorted().collect(Collectors.toList());
+    }
+    throw new IllegalArgumentException(
+        "Unexpected flattening config: " + formatConfig.getFlattening());
+  }
+
+  static String getOutputPrefix(String bucketPath, AirbyteStream stream) {
+    return getOutputPrefix(bucketPath, stream.getNamespace(), stream.getName());
+  }
+
+  @VisibleForTesting
+  public static String getOutputPrefix(String bucketPath, String namespace, String streamName) {
+    List<String> paths = new LinkedList<>();
+
+    if (bucketPath != null) {
+      paths.add(bucketPath);
+    }
+    if (namespace != null) {
+      paths.add(NAME_TRANSFORMER.convertStreamName(namespace));
+    }
+    paths.add(NAME_TRANSFORMER.convertStreamName(streamName));
+
+    return String.join("/", paths);
+  }
+
+  static String getOutputFilename(Timestamp timestamp) {
+    return String.format("%s-%d.csv", DATE_FORMAT.format(timestamp), timestamp.getTime());
```

Comment:
I wonder if we should add `%s-%d-1.csv` to anticipate partitioning down the line

Reply:
Done. The partition ID will start with `0`.